plugin/pkg/admission/unsafedeletepolicy/admission.go

```diff
@@ -0,0 +1,189 @@
+	auditReasonKey   = prefix + "reason"
+)
+
+// Register registers a plugin
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		return NewUnsafeDeletePolicy(), nil
+	})
+}
+
+// NewUnsafeDeletePolicy creates a new admission control handler that will admit
+// an unsafe delete of a corrupt object if the user has the appropriate privilege
+func NewUnsafeDeletePolicy() *UnsafeDeletePolicy {
+	return &UnsafeDeletePolicy{
+		Handler: admission.NewHandler(admission.Delete),
+	}
+}
+
+type UnsafeDeletePolicy struct {
+	*admission.Handler
+	authz authorizer.Authorizer
+}
+
+var _ admission.ValidationInterface = &UnsafeDeletePolicy{}
+var _ admission.InitializationValidator = &UnsafeDeletePolicy{}
+var _ genericadmissioninit.WantsAuthorizer = &UnsafeDeletePolicy{}
+
+// SetAuthorizer sets the authorizer.
+func (p *UnsafeDeletePolicy) SetAuthorizer(authz authorizer.Authorizer) {
+	p.authz = authz
+}
+
+// ValidateInitialization ensures an authorizer is set.
+func (p *UnsafeDeletePolicy) ValidateInitialization() error {
+	if p.authz == nil {
+		return fmt.Errorf("%s requires an authorizer", PluginName)
+	}
+	return nil
+}
+
+// Validate ensures that the user has permission to do 'delete-ignore-read-errors'
+// on the resource being deeted when ignoreStoreReadErrorWithClusterBreakingPotential
+// is enabled, these are the constraints:
+// a) must be a DELETE operation
+// b) ignoreStoreReadErrorWithClusterBreakingPotential is set to true
+// c) the request is a resource, no sub-resource or non-resource endpoint
+// d) the user has permission to do 'delete-ignore-read-errors' on the resource
+func (p *UnsafeDeletePolicy) Validate(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces) (err error) {
+	// if the feature is disabled, this plugin, although enabled should not
+	// be active, and requests with the delete option
+	// 'ignoreStoreReadErrorWithClusterBreakingPotential' enabled will
+	// fallback to using the normal deletion flow
+	if !utilfeature.DefaultFeatureGate.Enabled(genericfeatures.AllowUnsafeMalformedObjectDeletion) {
+		return nil
+	}
+	if attr.GetOperation() != admission.Delete || attr.GetOperationOptions() == nil {
+		return nil
+	}
+
+	options, ok := attr.GetOperationOptions().(*metav1.DeleteOptions)
+	if !ok {
+		return fmt.Errorf("expected an option of type: %T, but got: %T", &metav1.DeleteOptions{}, attr.GetOperationOptions())
+	}
+	if !ptr.Deref(options.IgnoreStoreReadErrorWithClusterBreakingPotential, false) {
+		return nil
+	}
+
+	requestInfo, found := request.RequestInfoFrom(ctx)
+	if !found {
+		return admission.NewForbidden(attr, errors.New("no RequestInfo found in the context"))
+	}
+	if !requestInfo.IsResourceRequest || len(attr.GetSubresource()) > 0 {
+		return admission.NewForbidden(attr, errors.New("ignoreStoreReadErrorWithClusterBreakingPotential delete option is not allowed on a subresource or non-resource request"))
+	}
```

Comment:
@deads2k any concern if we remove the admission plugin and do the access control inside the registry?

Reply:
> Why is this an admission plugin instead of just part of the REST layer?

I'm not aware of the REST or storage layer taking an authorizer.  Perhaps it was done for SSA.  If there's precedent and it's clean, pushing it a layer deeper (and well gated) seems ok.

Reply:
SSA patch does it here, i think
https://github.com/kubernetes/kubernetes/blob/50998de605d59bc02fe3ece4696d034e92a99e6a/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go#L697-L704

I will look into moving it to the REST layer cleanly.

Reply:
I found it here: https://github.com/kubernetes/kubernetes/blob/37ed906a33211c7d578cab2d681941ebfd2f2f23/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go#L232

yeah, pushing into endpoint/handlers is good.

Reply:
this has been moved to endpoint/handlers https://github.com/kubernetes/kubernetes/pull/127513/commits/95179fb759e4cdb99ecf12793a7a2d93e73f0034

Reply:
resolving it, please review the access control check inside endpoints/handlers